<a href="https://colab.research.google.com/github/sharat-shankar/Deep-Learning/blob/master/J045_Lab6_1_Keras_Activations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of activations

Activations can either be used through an Activation layer, or through the activation argument supported by all forward layers:



```
from keras.layers import Activation, Dense
model.add(Dense(64))
model.add(Activation('tanh'))
```



This is equivalent to:

```
model.add(Dense(64, activation='tanh'))
```

You can also pass an element-wise TensorFlow/Theano/CNTK function as an activation:

```
from keras import backend as K
model.add(Dense(64, activation=K.tanh))
```

#elu

keras.activations.elu(x, alpha=1.0)

Exponential linear unit.

##Arguments

1. x: Input tensor.
2. alpha: A scalar, slope of negative section.

Returns

1. The exponential linear activation: x if x > 0 and alpha * (exp(x)-1) if x < 0.

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


11493376/11490434 [==============================] - 1s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 11s 182us/step - loss: 0.2845 - acc: 0.9130 - val_loss: 0.1422 - val_acc: 0.9572
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1195 - acc: 0.9632 - val_loss: 0.1006 - val_acc: 0.9678
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0818 - acc: 0.9737 - val_loss: 0.0895 - val_acc: 0.9742
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0616 - acc: 0.9806 - val_loss: 0.0738 - val_acc: 0.9770
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0483 - acc: 0.9846 - val_loss: 0.0802 - val_acc: 0.9765
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0408 - acc

#softmax

keras.activations.softmax(x, axis=-1)

##Softmax activation function.

##Arguments

1. x: Input tensor.
2. axis: Integer, axis along which the softmax normalization is applied.

##Returns

Tensor, output of softmax transformation.

##Raises

ValueError: In case dim(x) == 1.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 2.2690 - acc: 0.1997 - val_loss: 2.1953 - val_acc: 0.6540
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 2.0402 - acc: 0.7452 - val_loss: 1.8487 - val_acc: 0.7568
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 1.6343 - acc: 0.7565 - val_loss: 1.4110 - val_acc: 0.7603
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 1.2140 - acc: 0.7648 - val_loss: 1.0252 - val_acc: 0.7615
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.8894 - acc: 0.7688 - val_loss: 0.7669 - val_acc: 0.7708
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.6931 - acc: 0.7748 - val_loss: 0.6291 - val_acc: 0.7801
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.5917 - acc: 0.7832 - val_loss: 0.5590 - val_acc

#selu

keras.activations.selu(x)

Scaled Exponential Linear Unit (SELU).

SELU is equal to: scale * elu(x, alpha), where alpha and scale are predefined constants. The values of alpha and scale are chosen so that the mean and variance of the inputs are preserved between two consecutive layers as long as the weights are initialized correctly (see lecun_normal initialization) and the number of inputs is "large enough" (see references for more information).

##Arguments

1. x: A tensor or variable to compute the activation function for.

##Returns

The scaled exponential unit activation: scale * elu(x, alpha).

##Note

1. To be used together with the initialization "lecun_normal".
2. To be used together with the dropout variant "AlphaDropout".

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.3316 - acc: 0.9017 - val_loss: 0.1438 - val_acc: 0.9563
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1372 - acc: 0.9586 - val_loss: 0.1240 - val_acc: 0.9611
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0929 - acc: 0.9711 - val_loss: 0.1410 - val_acc: 0.9580
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0725 - acc: 0.9768 - val_loss: 0.1145 - val_acc: 0.9685
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0603 - acc: 0.9808 - val_loss: 0.0814 - val_acc: 0.9763
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0504 - acc: 0.9843 - val_loss: 0.1078 - val_acc: 0.9713
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0417 - acc: 0.9866 - val_loss: 0.0738 - val_acc

#softplus

keras.activations.softplus(x)

Softplus activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softplus activation: log(exp(x) + 1).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.4495 - acc: 0.8618 - val_loss: 0.1965 - val_acc: 0.9390
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1691 - acc: 0.9484 - val_loss: 0.1786 - val_acc: 0.9462
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1121 - acc: 0.9656 - val_loss: 0.1185 - val_acc: 0.9625
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0831 - acc: 0.9747 - val_loss: 0.0841 - val_acc: 0.9751
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0649 - acc: 0.9801 - val_loss: 0.0927 - val_acc: 0.9726
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0525 - acc: 0.9833 - val_loss: 0.0787 - val_acc: 0.9761
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0437 - acc: 0.9857 - val_loss: 0.0750 - val_acc

#softsign

keras.activations.softsign(x)

##Softsign activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softsign activation: x / (abs(x) + 1).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.2923 - acc: 0.9140 - val_loss: 0.1720 - val_acc: 0.9466
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1462 - acc: 0.9566 - val_loss: 0.1146 - val_acc: 0.9658
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1011 - acc: 0.9691 - val_loss: 0.0968 - val_acc: 0.9704
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0763 - acc: 0.9768 - val_loss: 0.0907 - val_acc: 0.9704
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0585 - acc: 0.9825 - val_loss: 0.0830 - val_acc: 0.9737
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0452 - acc: 0.9860 - val_loss: 0.0685 - val_acc: 0.9780
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0358 - acc: 0.9888 - val_loss: 0.0749 - val_acc

#relu

keras.activations.relu(x, alpha=0.0, max_value=None, threshold=0.0)

##Rectified Linear Unit.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = alpha * (x - threshold) otherwise.

##Arguments

1. x: Input tensor.
2. alpha: float. Slope of the negative part. Defaults to zero.
3. max_value: float. Saturation threshold.
4. threshold: float. Threshold value for thresholded activation.

##Returns

A tensor.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.2294 - acc: 0.9298 - val_loss: 0.1178 - val_acc: 0.9626
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0847 - acc: 0.9734 - val_loss: 0.0948 - val_acc: 0.9712
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0561 - acc: 0.9822 - val_loss: 0.0958 - val_acc: 0.9710
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0389 - acc: 0.9879 - val_loss: 0.0695 - val_acc: 0.9805
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0310 - acc: 0.9905 - val_loss: 0.0806 - val_acc: 0.9801
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0209 - acc: 0.9933 - val_loss: 0.0809 - val_acc: 0.9801
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0186 - acc: 0.9947 - val_loss: 0.0847 - val_acc

#tanh

keras.activations.tanh(x)

##Hyperbolic tangent activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The hyperbolic activation: tanh(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x))

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.2935 - acc: 0.9117 - val_loss: 0.1620 - val_acc: 0.9504
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1246 - acc: 0.9622 - val_loss: 0.1077 - val_acc: 0.9669
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0814 - acc: 0.9755 - val_loss: 0.0870 - val_acc: 0.9732
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0585 - acc: 0.9824 - val_loss: 0.0700 - val_acc: 0.9783
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0434 - acc: 0.9869 - val_loss: 0.0747 - val_acc: 0.9756
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0320 - acc: 0.9901 - val_loss: 0.0834 - val_acc: 0.9721
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0236 - acc: 0.9930 - val_loss: 0.0681 - val_acc

#sigmoid

keras.activations.sigmoid(x)

##Sigmoid activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The sigmoid activation: 1 / (1 + exp(-x)).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.4911 - acc: 0.8593 - val_loss: 0.2555 - val_acc: 0.9234
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.2170 - acc: 0.9346 - val_loss: 0.1818 - val_acc: 0.9438
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1572 - acc: 0.9522 - val_loss: 0.1403 - val_acc: 0.9576
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1195 - acc: 0.9637 - val_loss: 0.1179 - val_acc: 0.9642
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0943 - acc: 0.9711 - val_loss: 0.1044 - val_acc: 0.9682
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0764 - acc: 0.9766 - val_loss: 0.0974 - val_acc: 0.9682
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0625 - acc: 0.9808 - val_loss: 0.0828 - val_acc

#hard_sigmoid

keras.activations.hard_sigmoid(x)

##Hard sigmoid activation function.

Faster to compute than sigmoid activation.

##Arguments

1. x: Input tensor.

##Returns

1. Hard sigmoid activation:

```
0 if x < -2.5
1 if x > 2.5
0.2 * x + 0.5 if -2.5 <= x <= 2.5.
```

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.4968 - acc: 0.8545 - val_loss: 0.2394 - val_acc: 0.9278
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.2096 - acc: 0.9372 - val_loss: 0.1770 - val_acc: 0.9477
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1478 - acc: 0.9564 - val_loss: 0.1286 - val_acc: 0.9613
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1105 - acc: 0.9671 - val_loss: 0.1092 - val_acc: 0.9666
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0851 - acc: 0.9742 - val_loss: 0.0976 - val_acc: 0.9708
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0681 - acc: 0.9790 - val_loss: 0.0885 - val_acc: 0.9725
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0557 - acc: 0.9834 - val_loss: 0.0784 - val_acc

#exponential

keras.activations.exponential(x)

Exponential (base e) activation function.

##Arguments

1. x: Input tensor.

##Returns

1. Exponential activation: exp(x).

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 14.2959 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 14.3070 - acc: 0.1124 - val_loss: 14.

#linear

keras.activations.linear(x)

Linear (i.e. identity) activation function.

##Arguments

1. x: Input tensor.

##Returns

Input tensor, unchanged.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.4240 - acc: 0.8761 - val_loss: 0.3345 - val_acc: 0.9067
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.3419 - acc: 0.9026 - val_loss: 0.3216 - val_acc: 0.9113
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.3227 - acc: 0.9084 - val_loss: 0.3042 - val_acc: 0.9150
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3096 - acc: 0.9143 - val_loss: 0.2972 - val_acc: 0.9172
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3018 - acc: 0.9159 - val_loss: 0.3061 - val_acc: 0.9156
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.2942 - acc: 0.9175 - val_loss: 0.3170 - val_acc: 0.9110
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2885 - acc: 0.9186 - val_loss: 0.2942 - val_acc

#LeakyReLU

keras.layers.LeakyReLU(alpha=0.3)

Leaky version of a Rectified Linear Unit.

It allows a small gradient when the unit is not active: 
```
f(x) = alpha * x for x < 0
f(x) = x for x >= 0.
```

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

alpha: float >= 0. Negative slope coefficient.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.2643 - acc: 0.9206 - val_loss: 0.1480 - val_acc: 0.9548
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1249 - acc: 0.9621 - val_loss: 0.1301 - val_acc: 0.9574
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0907 - acc: 0.9720 - val_loss: 0.0974 - val_acc: 0.9708
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0741 - acc: 0.9763 - val_loss: 0.0862 - val_acc: 0.9736
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0630 - acc: 0.9803 - val_loss: 0.1116 - val_acc: 0.9700
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0540 - acc: 0.9826 - val_loss: 0.0947 - val_acc: 0.9726
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0472 - acc: 0.9854 - val_loss: 0.0844 - val_acc

#PReLU

keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)

Parametric Rectified Linear Unit.

It follows: f(x) = alpha * x for x < 0, f(x) = x for x >= 0, where alpha is a learned array with the same shape as x.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha_initializer: initializer function for the weights.
2. alpha_regularizer: regularizer for the weights.
3. alpha_constraint: constraint for the weights.
4. shared_axes: the axes along which to share learnable parameters for the activation function. For example, if the incoming feature maps are from a 2D convolution with output shape (batch, height, width, channels), and you wish to share parameters across space so that each filter only has one set of parameters, set shared_axes=[1, 2].

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.2258 - acc: 0.9304 - val_loss: 0.1353 - val_acc: 0.9585
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0831 - acc: 0.9744 - val_loss: 0.0873 - val_acc: 0.9724
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0549 - acc: 0.9829 - val_loss: 0.0817 - val_acc: 0.9777
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0400 - acc: 0.9875 - val_loss: 0.0732 - val_acc: 0.9804
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0295 - acc: 0.9905 - val_loss: 0.0663 - val_acc: 0.9824
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0230 - acc: 0.9928 - val_loss: 0.0687 - val_acc: 0.9826
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0170 - acc: 0.9947 - val_loss: 0.0813 - val_acc

#ELU

keras.layers.ELU(alpha=1.0)

Exponential Linear Unit.

It follows: f(x) =  alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha: scale for the negative factor.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.2904 - acc: 0.9116 - val_loss: 0.1378 - val_acc: 0.9572
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1201 - acc: 0.9626 - val_loss: 0.1094 - val_acc: 0.9658
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0828 - acc: 0.9740 - val_loss: 0.0857 - val_acc: 0.9729
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0629 - acc: 0.9803 - val_loss: 0.0840 - val_acc: 0.9747
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0499 - acc: 0.9840 - val_loss: 0.0849 - val_acc: 0.9734
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0393 - acc: 0.9870 - val_loss: 0.0725 - val_acc: 0.9786
Epoch 7/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0331 - acc: 0.9890 - val_loss: 0.0807 - val_acc

#ThresholdedReLU

keras.layers.ThresholdedReLU(theta=1.0)

Thresholded Rectified Linear Unit.

It follows: f(x) = x for x > theta, f(x) = 0 otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. theta: float >= 0. Threshold location of activation.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.7102 - acc: 0.7577 - val_loss: 0.1905 - val_acc: 0.9403
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1569 - acc: 0.9525 - val_loss: 0.1181 - val_acc: 0.9657
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1021 - acc: 0.9685 - val_loss: 0.0963 - val_acc: 0.9700
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0790 - acc: 0.9754 - val_loss: 0.0907 - val_acc: 0.9726
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0614 - acc: 0.9810 - val_loss: 0.0838 - val_acc: 0.9746
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0509 - acc: 0.9839 - val_loss: 0.0908 - val_acc: 0.9748
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0410 - acc: 0.9877 - val_loss: 0.0845 - val_acc

#Softmax

keras.layers.Softmax(axis=-1)

Softmax activation function.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. axis: Integer, axis along which the softmax normalization is applied.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 2.2701 - acc: 0.1997 - val_loss: 2.1994 - val_acc: 0.6527
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 2.0482 - acc: 0.6929 - val_loss: 1.8620 - val_acc: 0.7448
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.6504 - acc: 0.7530 - val_loss: 1.4291 - val_acc: 0.7586
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 1.2360 - acc: 0.7616 - val_loss: 1.0498 - val_acc: 0.7611
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.9111 - acc: 0.7668 - val_loss: 0.7870 - val_acc: 0.7717
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.7078 - acc: 0.7757 - val_loss: 0.6420 - val_acc: 0.7767
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.5990 - acc: 0.7825 - val_loss: 0.5704 - val_acc

#ReLU

keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)

Rectified Linear Unit activation function.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = negative_slope * (x - threshold) otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. max_value: float >= 0. Maximum activation value.
2. negative_slope: float >= 0. Negative slope coefficient.
3. threshold: float. Threshold value for thresholded activation.

In [0]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2297 - acc: 0.9290 - val_loss: 0.1242 - val_acc: 0.9599
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0837 - acc: 0.9744 - val_loss: 0.0755 - val_acc: 0.9767
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0559 - acc: 0.9829 - val_loss: 0.0679 - val_acc: 0.9810
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0389 - acc: 0.9875 - val_loss: 0.0835 - val_acc: 0.9754
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0295 - acc: 0.9907 - val_loss: 0.0725 - val_acc: 0.9797
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0224 - acc: 0.9931 - val_loss: 0.0802 - val_acc: 0.9808
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0176 - acc: 0.9945 - val_loss: 0.0844 - val_acc

#Thank you for completing this notebook

ReLU is the best activation with maximum accuracy of 98.29% and loss 0.0920575719

Fasion MNIST

In [0]:
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.layers import Activation, Flatten, Conv2D

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))
X_test = X_test.astype('float32') / 255



# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)

# Build neural network
model = models.Sequential()
model.add(Conv2D(512, (28, 28), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 126us/step - loss: 0.6029 - acc: 0.7744 - val_loss: 0.4781 - val_acc: 0.8241
Epoch 2/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.3995 - acc: 0.8517 - val_loss: 0.4609 - val_acc: 0.8381
Epoch 3/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.3539 - acc: 0.8689 - val_loss: 0.4472 - val_acc: 0.8340
Epoch 4/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.3288 - acc: 0.8777 - val_loss: 0.3677 - val_acc: 0.8739
Epoch 5/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.3103 - acc: 0.8854 - val_loss: 0.4043 - val_acc: 0.8761
Epoch 6/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2970 - acc: 0.8911 - val_loss: 0.3607 - val_acc: 0.8806
Epoch 7/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2860 - acc: 0.8947 - val_loss: 0.3825 - val_ac

CIFAR10

In [0]:
from keras.datasets import cifar10
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

170500096/170498071 [==============================] - 4s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 6s 112us/step - loss: 2.0332 - acc: 0.2947 - val_loss: 1.8095 - val_acc: 0.3351
Epoch 2/10
50000/50000 [==============================] - 5s 90us/step - loss: 1.7195 - acc: 0.3828 - val_loss: 1.7096 - val_acc: 0.3905
Epoch 3/10
50000/50000 [==============================] - 4s 88us/step - loss: 1.6246 - acc: 0.4199 - val_loss: 1.5651 - val_acc: 0.4496
Epoch 4/10
50000/50000 [==============================] - 4s 86us/step - loss: 1.5552 - acc: 0.4443 - val_loss: 1.5494 - val_acc: 0.4415
Epoch 5/10
50000/50000 [==============================] - 4s 86us/step - loss: 1.5014 - acc: 0.4639 - val_loss: 1.5009 - val_acc: 0.4701
Epoch 6/10
50000/50000 [==============================] - 4s 87us/step - loss: 1.4523 - acc: 0.4808 - val_loss: 1.5086 - val_acc: 0.4610
Epoch 7/10
50000/50000 [==============================] - 4s 89us/step - loss: 1.4215 - acc: 0.4930 - val_loss: 1.4738 - val_ac

CIFAR 100

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar100
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(100, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 4s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 5s 109us/step - loss: 4.1915 - acc: 0.0632 - val_loss: 3.9326 - val_acc: 0.0938
Epoch 2/10
50000/50000 [==============================] - 4s 87us/step - loss: 3.7680 - acc: 0.1253 - val_loss: 3.6825 - val_acc: 0.1453
Epoch 3/10
50000/50000 [==============================] - 4s 87us/step - loss: 3.5787 - acc: 0.1563 - val_loss: 3.5435 - val_acc: 0.1714
Epoch 4/10
50000/50000 [==============================] - 4s 89us/step - loss: 3.4427 - acc: 0.1821 - val_loss: 3.4779 - val_acc: 0.1772
Epoch 5/10
50000/50000 [==============================] - 4s 87us/step - loss: 3.3380 - acc: 0.2006 - val_loss: 3.3978 - val_acc: 0.1972
Epoch 6/10
50000/50000 [==============================] - 4s 87us/step - loss: 3.2464 - acc: 0.2167 - val_loss: 3.3711 - val_acc: 0.2036
Epoch 7/10
50000/50000 [==============================] - 4s 86us/step - loss: 3.1596 - acc: 0.2301 - val_loss: 3.3020 - val_ac